### ISSUE TITLE
Oracle review comments on CR draft - Part 1

### ISSUE THREAD

**Issue by mlagally** [w3c/wot-thing-description#1548]
# Oracle review comments (Part 1)

## RFC2119 assertions

There are multiple requriements in normative sections which contain an optional clause (should, may, recommended) in lower case. Several of them are important assertions and should be marked up with RFC2119 markup.

Examples include the language around TD processors and TD producers in *5.3.1.1 Thing*, content of some tables in chapter *5.3*, e.g. *5.3.3.9 OAuth2SecurityScheme* and several clauses in *8. Behavioral Assertions*.

## Hints
In several sections, the text talks about hints, e.g. *5.3.4.1 Link*: 

type: "Target attribute providing a hint indicating what the media type [RFC2046] of the result of dereferencing the link should be."

To achieve interoperability this is not sufficient. it should be an assertion that only media type values defined by IANA *MUST* be used.
 

### 5.3 Class Definitions
The section should make it a mandatory requirement that vocabulary **terms** MUST be case sensitive.

#### 5.3.1.1 Thing
- security: "Set of security definition names, chosen from those defined in securityDefinitions. These must all be satisfied for access to resources." What does "all be satisfied" mean? Bettwe reword to "security mechanisms supported by the thing"

##### 5.3.1.2 InteractionAffordance
- uriVariables: ObjectSchema and ArraySchema are forbidden, however there's an assertion required.
- forms: forms are mandatory, need to clarify if 0-element forms are permitted.

##### 5.3.1.3 PropertyAffordance
- invalid combinations of op values: some combinations don't make sense and should be prohibited. Examples include observe+unobserve.
- observable: RFC2119 *SHOULD* assertion

##### 5.3.1.4 ActionAffordance
- combinations of operations do not make sense and should not be permitted.

- An operation *cancelAllActions* would make sense as an emergency stop feature.

##### 5.3.1.5 EventAffordance
- data: is currently optional, should be mandatory since an event subscription without a data flow does not make sense.
- combinations of operations do not make sense and should not be permitted.

##### 5.3.1.6 VersionInfo

- "It is recommended that the values within instances and model of the VersionInfo Class follow the semantic versioning pattern ..." should be an RFC2119 assertion.

#### 5.3.2 Data Schema Vocabulary Definitions

- The following should be an RFC2119 assertion: "it is noted that data schema definitions within Thing Description instances are not limited to this defined subset and MAY use additional terms found in JSON Schema". 
- What does it mean "the table is at risk"? all entries?

##### 5.3.2.1 DataSchema

- enum: is defined as an array of *any type* without details. *type* lists possible values for any type (including null). Is null valid for enum?
- default: missing RFC2119 markup. Consider **MUST**.

##### 5.3.2.2 ArraySchema

- why are items optional? Does it make sense to define an array without items?

##### 5.3.2.6 ObjectSchema

- required: what does "members are mandatory" imply?

##### 5.3.2.7 StringSchema

- contentEncoding: please add link to RFC
- contentMediaType: please add link to RFC
- RFC2119 assertion: "Since this value is not a fragment identifier, and is not relative to the root of the TD but to whichever data schemas the security scheme is bound to, this value SHOULD not start with #;"

#### 5.3.3.1 SecurityScheme

- "For all security schemes, any private keys, passwords, or other sensitive information directly providing access should be shared and stored out-of-band and MUST NOT be stored in the TD." break down into two sentences. First *should*

- @type: semantic tags and semantic types should be defined in the terminology.
- **default language** - consider recommending english.
- support: "https" seems to be a left over.
- uriVariables: precedence rules for resolving name conflicts are harmful. why not requre that all uriVariable definitions are unique within a TD?
- uriVariables: The following sentence is a bit confusing: "The individual variables DataSchema cannot be an ObjectSchema or an ArraySchema."
- In symmetry with queryallactions and unsubscribeallevents at Thing level, an operation "cancelallactions" should be defined, to enable a simple emergency stop.
- writemultiple has difficult semantics, how about read only? Best effort approach? Fail, if one of the writes is failing?

#### 5.3.3.9 OAuth2SecurityScheme

- scopes: RFC2119 assertion: "The values associated with a form should be chosen from those defined in an OAuth2SecurityScheme active on that form."

#### 5.3.1.6 VersionInfo
- semantic versioning should be **RECOMMENDED**
- typo: (unnecessary ".") "underlying TM. instance"

#### 5.3.4.2 Form

The concept of ExpectedResponse and AdditionalExpectedResponse needs more clarification. 

There is an assumption that there is a "primary" response and a set of "error" responses. However there may be multiple primary responses with different content types - an example would be an action that returns an image in different image formats (svg, png, jpg, ...) All of these are primary responses. 

One way to resolve the issue is to define the response as an set of responses and include the fields from *AdditionalExpectedResponse* to that *Response* type.

#### 6.3.4 securityDefinitions and security

- missing normative assertions: 

"A named security configuration must be given in the securityDefinitions map. That definition must be activated by including its JSON name in the security member, which can and should be of type string when only one definition is activated."

### 11. Thing Model

- The note states that the section is work in progress. Will this note be removed? What is currently missing from the draft?

If this chapter is not stable or complete, it should be made **informative**.

- normative assertion? : "This kind of template should be valid and followed for all instantiated Thing Descriptions that are relied on this Thing Model."

#### 11.3.1 Versioning

- normative assertion: "Over time, Thing Model definitions may change and must be made identifiable through versioning."

**Comment by mmccool** [comment]
To organize: checkboxes, or smaller issues for each point?
Conclusion: use checkboxes; Sebastian to review and generate the list of actionable items and add them to this issue as a comment.

**Comment by egekorkan** [comment]
@mlagally below are my comments to parts that need discussion. We should open separate issues for those if you think that my explanation is not enough.

> ## Hints
> 
> In several sections, the text talks about hints, e.g. _5.3.4.1 Link_:
> 
> type: "Target attribute providing a hint indicating what the media type [RFC2046] of the result of dereferencing the link should be."
> 
> To achieve interoperability this is not sufficient. it should be an assertion that only media type values defined by IANA _MUST_ be used.

We define our own values as well so we cannot limit to IANA only. However, it would be a valid point to discuss whether we should limit it to what IANA defines + what we define. In any case, restricting these values should be done in TD2.0 since TD1.0 TDs might be using some custom values or annotated ones.

> #### 5.3.1.1 Thing
> 
>     * security: "Set of security definition names, chosen from those defined in securityDefinitions. These must all be satisfied for access to resources." What does "all be satisfied" mean? Bettwe reword to "security mechanisms supported by the thing"

I would say we should reword it to "the credentials corresponding to the security definitions MUST be all provided during the operation"
> 
> ##### 5.3.1.2 InteractionAffordance
> 
>     * uriVariables: ObjectSchema and ArraySchema are forbidden, however there's an assertion required.

I sort of agree. However, everything inside the tables are assertive. 

>     * forms: forms are mandatory, need to clarify if 0-element forms are permitted.
> 

I agree. It is done in [the JSON Schema](https://github.com/w3c/wot-thing-description/blob/main/validation/td-json-schema-validation.json#L657) but not specified in the text


> ##### 5.3.1.3 PropertyAffordance
> 
>     * invalid combinations of op values: some combinations don't make sense and should be prohibited. Examples include observe+unobserve.

Actually, exactly this makes sense. In MQTT, we have the default assumption that observing is subscribing and unobserving is unsubscribing on the same topic. Thus, in MQTT it would redundant to split the operations. This is also valid for events.

> ##### 5.3.1.4 ActionAffordance
> 
>     * combinations of operations do not make sense and should not be permitted.

Might make sense but a protocol's defaults can allow this to be done.

> 
>     * An operation _cancelAllActions_ would make sense as an emergency stop feature.
> 

Should be deferred to TD2.0 at this stage.

> ##### 5.3.1.5 EventAffordance
> 
>     * data: is currently optional, should be mandatory since an event subscription without a data flow does not make sense.
> 

Not necessarily. If an event is called `buttonPressed`, it does not even need a payload.

> ##### 5.3.1.6 VersionInfo
> 
>     * "It is recommended that the values within instances and model of the VersionInfo Class follow the semantic versioning pattern ..." should be an RFC2119 assertion.
> 

I agree but semantic versioning and its meaning in the context of TDs is not defined. I would say that even the working group would give different answers to what means a PATCH change to a TD. I think that if we are putting this as an assertion, we must define what PATCH, MINOR and MAJOR mean in the case of TDs.

> ##### 5.3.2.1 DataSchema
> 
>     * enum: is defined as an array of _any type_ without details. _type_ lists possible values for any type (including null). Is null valid for enum?

Yes.


> ##### 5.3.2.2 ArraySchema
> 
>     * why are items optional? Does it make sense to define an array without items?

This is just a copy over from JSON Schema where `items `term is optional. Probably the reason is that there is also the `prefixItems` term that can be used for tuple validation. See https://json-schema.org/understanding-json-schema/reference/array.html#id

> ##### 5.3.2.6 ObjectSchema
> 
>     * required: what does "members are mandatory" imply?
> 

This definitely needs an improvement. It means what members are mandatory in the payload that is to be sent (e.g. `invokeaction` input, `writeproperty`) and what members will be definitely delivered in the payload that is being received (e.g. `invokeaction` output, `readproperty`)


>     * **default language** - consider recommending english.

I think we should not do it to avoid bad internationalization reviews. 

>     * support: "https" seems to be a left over.
> 
>     * uriVariables: precedence rules for resolving name conflicts are harmful. why not requre that all uriVariable definitions are unique within a TD?

Although it makes sense, since TD is generic we should be able to adapt to this situation where the name is used twice with different parameters. 

>     * In symmetry with queryallactions and unsubscribeallevents at Thing level, an operation "cancelallactions" should be defined, to enable a simple emergency stop.

Too late at this point

>     * writemultiple has difficult semantics, how about read only? Best effort approach? Fail, if one of the writes is failing?

I agree, however, there was no proper discussion on this. I could not find the issue.

>     * semantic versioning should be **RECOMMENDED**

See my point above about semver

**Comment by egekorkan** [comment]
Call of 13.07:
- We would like to wait @mlagally to give his opinion in one of the next TD calls

**Comment by mlagally** [comment]
@egekorkan 

> To organize: checkboxes, or smaller issues for each point? Conclusion: use checkboxes; Sebastian to review and generate the list of actionable items and add them to this issue as a comment.

I'm waiting for this break-down into individual issues. 
Discussing these in one big issue will lead to a lot of confusion and misunderstandings, and is likely not productive.
This also allows for prioritisation. 

I will attend one of the next TD calls (unfortunately I can only participate in the first hour) to discuss individual items.

**Comment by egekorkan** [comment]
@mlagally I have opened issues 1599 to 1619 with your comments. Each issue is addressing one title you have used. Thus, some issues have actually multiple discussion issues. If this granularity is not enough, we should open further granular issues. Simply click the circle next to the checkbox text to open an issue. 

In any case, I propose closing this issue and continuing the discussion there

**Comment by mlagally** [comment]
@egekorkan 
Thanks for splitting into smaller issues. I'm fine with closing this issue.
I hope to find time for Part 2 of the review soon.

**Comment by egekorkan** [comment]
thank **you** for taking the time to review the spec :)
